File: module_3/python/has_cycle.py
"""
Title: Does linked list have cycle.

Problem:
        Given a linked list, determine if it has a cycle in it.

        To represent a cycle in the given linked list, we use an integer pos which
        represents the position (0-indexed) in the linked list where tail connects to.
        If pos is -1, then there is no cycle in the linked list.

Execution: python has_cycle.py
"""
import unittest


class Node:
    """Node class for LinkedList."""

    def __init__(self, data):
        """Set the data for node."""
        self.data = data
        self.next = None


class LinkedList:
    """Linked list class."""

    def __init__(self): 
        """Initialize head of list."""
        self.head = None

    def append(self, new_data): 
        """Insert a new node at the beginning."""
        new_node = Node(new_data)
        new_node.next = self.head
        self.head = new_node

    def print_list(self):
        """Print the linked LinkedList."""
        temp = self.head
        while(temp):
            print(temp.data)
            temp = temp.next

    def detect_loop(self):
        """Check if LinkedList has loop."""
        s = set()
        temp = self.head
        while (temp):
            # If we have already has this node in hashmap it
            # means their is a cycle (Because you we encountering
            # the node second time).
            if (temp in s):
                return True

            # If we are seeing the node for
            # the first time, insert it in hash
            s.add(temp)
            temp = temp.next

        return False


class TestHasCycle(unittest.TestCase):
    """Unit test for has cycle function."""

    def test_1(self):
        ll = LinkedList()
        ll.append(20)
        ll.append(4)
        ll.append(15)
        ll.append(10)
        ll.head.next.next.next.next = ll.head
        self.assertEqual(ll.detect_loop(), True)


if __name__ == '__main__':
    unittest.main()

File: module_3/java/HasCycle.java
/*
 *   Title: Has cycle
 *
 *   Problem:
 *      Given a linked list, determine if it has a cycle in it.
 *
 *      To represent a cycle in the given linked list, we use an integer pos which
 *      represents the position (0-indexed) in the linked list where tail connects to.
 *      If pos is -1, then there is no cycle in the linked list.
 *
 *   Code is adapted from https://geeksforgeeks.org/detect-loop-in-a-linked-list/
 *
 *   Execution: javac HasCycle.java && java HasCycle
 */
import java.util.*;


public class HasCycle {
    
    public static class LinkedList {

        static Node head; // head of list

        static class Node {
            int data;
            Node next;
            Node(int d) {
                data = d;
                next = null;
            }
        }

        /* Inserts a new Node at front of the list. */
        public static void push(int new_data) {
            /* 1 & 2: Allocate the Node &
                      Put in the data*/
            Node new_node = new Node(new_data);

            /* 3. Make next of new Node as head */
            new_node.next = head;

            /* 4. Move the head to point to new Node */
            head = new_node;
        }

        // Returns true if there is a loop in linked
        // list else returns false.
        public static boolean hasCycle(Node h) {
            HashSet<Node> s = new HashSet<Node>();
            while (h != null) {
                // If we have already has this node
                // in hashmap it means their is a cycle
                // (Because you we encountering the
                // node second time).
                if (s.contains(h))
                    return true;

                // If we are seeing the node for
                // the first time, insert it in hash
                s.add(h);

                h = h.next;
            }

            return false;
        }
    }

    public static void main(String[] args) {
        LinkedList llist = new LinkedList(); 
  
        llist.push(20); 
        llist.push(4); 
        llist.push(15); 
        llist.push(10); 
  
        llist.head.next.next.next.next = llist.head; 

        assert llist.hasCycle(llist.head) == true;
        System.out.println("Explanation: ");

        System.out.println("Passed all test cases");
    }
    
}
